collections/courses.js

```diff
@@ -320,16 +349,30 @@ if (Meteor.isServer) {
 			Courses.update(courseId, { $set: {time_lastedit: new Date()} });
 		},
 
-		remove_role: function(courseId, role) {
+		remove_role: function(courseId, userId, role) {
 			check(role, String);
+			check(userId, String);
 			check(courseId, String);
 
-			var user = Meteor.user();
-			if (!user) throw new Meteor.Error(401, "please log in");
+			var user = Meteor.users.findOne(userId);
+			if (!user) throw new Meteor.Error(404, "User not found");
+
+			var operator = Meteor.user();
+			if (!operator) throw new Meteor.Error(401, "please log in");
 
 			var course = Courses.findOne({_id: courseId});
 			if (!course) throw new Meteor.Error(404, "Course not found");
 
+			// do nothing if user is not subscribed with this role
+			if (!hasRoleUser(course.members, role, userId)) return true;
+
+			// Check permissions
+			 if (!mayUnSubscribe(operator._id, course, user._id, role)) {
+				throw new Meteor.Error(401, "not permitted");
+			}
+
+			//what to do for incognito unsubscribe???
```

Comment:
Good thing you didn't forget about those. I would have thought that code was fine but once again it's the incognito to complicate things.  We fail too early above because we insist that the passed `userId` must be present in the `users` collection. The call to `hasRoleUser()` would actually return 'subscribed' even for anonId. `mayUnsubscribe()` works fine because it uses `hasRoleUser()` internally.

So what needs to be done here is to try and load the user by anonId as in
 ```
var user = Meteor.users.findOne(userId);
if (!user) {
    user = Meteor.users.findOne({anonId: userId});
    if (!user) throw new Meteor.Error(404, "User not found");
}
```
Now the problem with this is that `mayUnsubscribe()` only allows removing roles for users themselves. So it would have to be made aware of the distinction too. I'd prefer getting rid of this feature altogether. To this end, I've created issue #633. Maybe we wait until we're done discussing that issue?